docs/book/src/developer/providers/contracts/infra-machine.md

```diff
@@ -1,118 +1,486 @@
+- `status`
+- `reason` ((optional, if omitted, a default one will be used)
+- `message` (optional)
+- `lastTransitionTime` (optional, if omitted, time.Now will be used)
+
+Other fields will be ignored
+
+See [Improving status in CAPI resources] for more context.
+
+Please note that provider that will continue to use legacy Cluster API condition types MUST carefully take into account
+the implication of this choice which are described both in the document above and in the notice at the beginning of the [Cluster API condition proposal]..
+
+</aside>
+
+### InfraMachine: terminal failures
+
+Each InfraMachine SHOULD report when Machine's enter in a state that cannot be recovered (terminal failure) by
+setting `status.failureReason` and `status.failureMessage` in the InfraMachine resource.
+
+```go
+type FooMachineStatus struct {
+    // FailureReason will be set in the event that there is a terminal problem reconciling the FooMachine 
+    // and will contain a succinct value suitable for machine interpretation.
+    //
+    // This field should not be set for transitive errors that can be fixed automatically or with manual intervention,
+    // but instead indicate that something is fundamentally wrong with the FooMachine and that it cannot be recovered.
+    // +optional
+    FailureReason *capierrors.ClusterStatusError `json:"failureReason,omitempty"`
+    
+    // FailureMessage will be set in the event that there is a terminal problem reconciling the FooMachine
+    // and will contain a more verbose string suitable for logging and human consumption.
+    //
+    // This field should not be set for transitive errors that can be fixed automatically or with manual intervention,
+    // but instead indicate that something is fundamentally wrong with the FooMachine and that it cannot be recovered.
+    // +optional
+    FailureMessage *string `json:"failureMessage,omitempty"`
+    
+    // See other rules for more details about mandatory/optional fields in InfraMachine status.
+    // Other fields SHOULD be added based on the needs of your provider.
+}
+```
+
+Once `status.failureReason` and `status.failureMessage` are set on the InfraMachine resource, the Machine "core" controller
+will bubble up those info in the corresponding fields in Machine's `status`.
+
+Please note that once failureReason/failureMessage is set in Machine's `status`, the only way to recover is to delete and
+recreate the Machine (it is a terminal failure).
+
+<aside class="note warning">
+
+<h1>Heads up! this will change with the v1beta2 contract</h1>
+
+When the v1beta2 contract will be released (tentative Apr 2025), support for `status.failureReason` and `status.failureMessage`
+will be dropped.
+
+See [Improving status in CAPI resources].
+
+</aside>
+
+### InfraMachineTemplate, InfraMachineTemplateList resource definition
+
+For a given InfraMachine resource, you should also add a corresponding InfraMachineTemplate resources in order to use it in ClusterClasses.
```

Comment:
I would simply drop the ClusterClass part. I would also consider changing this to MUST. Almost nothing works without InfraMachineTemplates

(or at least its must after folks implemented an InfraMachine)